common/src/main/java/io/netty/util/concurrent/PromiseOperator.java

```diff
@@ -0,0 +1,132 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import static io.netty.util.internal.PromiseNotificationUtil.tryCancel;
+import static io.netty.util.internal.PromiseNotificationUtil.tryFailure;
+import static io.netty.util.internal.PromiseNotificationUtil.trySuccess;
+
+/**
+ * Internal operators interface used for implementing {@link Future#map(Function)}
+ * and {@link Future#flatMap(Function)}.
+ * <p>
+ * <em>Note:</em> The operations themselves are implemented as static inner classes instead of lambdas to aid debugging.
+ * <p>
+ * This library reduces object allocation compared to what could otherwise be accomplished, by relying on the
+ * {@link DefaultPromise} itself being a {@link FutureContextListener} that take the operator instance as a context.
+ * This way, only the operator instance itself necessarily needs to be allocated, and in the case of
+ * {@link #passThrough()} even that is avoided.
+ */
+interface PromiseOperator<R> {
+    InternalLogger LOGGER = InternalLoggerFactory.getInstance(PromiseOperator.class);
+    PassThrough<?> PASS_THROUGH_CONSTANT = new PassThrough<>();
+
+    @SuppressWarnings("unchecked")
+    static <R> PromiseOperator<R> passThrough() {
+        return (PromiseOperator<R>) PASS_THROUGH_CONSTANT;
+    }
+
+    static <T, R> PromiseOperator<R> map(Function<T, R> mapper) {
+        Objects.requireNonNull(mapper, "The mapper function cannot be null.");
+         return new Mapper<>(mapper);
+    }
+
+    static <T, R> PromiseOperator<R> flatMap(Function<T, Future<R>> mapper) {
+        return new FlatMapper<>(mapper);
+    }
+
+    static <A, B> void propagateUncommonCompletion(Future<? extends A> completed, Promise<B> recipient) {
+        if (completed.isCancelled()) {
+            tryCancel(recipient, LOGGER);
+        } else {
+            Throwable cause = completed.cause();
+            tryFailure(recipient, cause, LOGGER);
+        }
+    }
+
+    void cascadeCompletion(Future<?> completed, DefaultPromise<R> recipient);
+
+    class PassThrough<R> implements PromiseOperator<R> {
+        @Override
+        public void cascadeCompletion(Future<?> completed, DefaultPromise<R> recipient) {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    R result = (R) completed.getNow();
+                    trySuccess(recipient, result, LOGGER);
```

Comment:
I think we should use `set*()` here as we own the promise and hence we should be the only one completing it. If we do this, we can call `setUncancellable()` first as cancellations are expected out of the context of this operator.

Reply:
Hmm… at least I don't think we should log anything if the futures got cancelled. In this particular case we can call `trySuccess` directly on the recipient. For the others, we can use `setUncancellable` to ensure the mapping function is only called when we know we're going to complete the future for sure.